write.go

```diff
@@ -232,12 +242,173 @@ func writeProduceRequestV2(w *bufio.Writer, codec CompressionCodec, correlationI
 	writeArrayLen(w, 1)
 	writeInt32(w, partition)
 	writeInt32(w, size)
+	w.Write(msgBuf)
+	return w.Flush()
+}
+
+func writeMessageSet(codec CompressionCodec, correlationID int32, clientId, topic string, partition int32, timeout time.Duration, requiredAcks int16, msgs ...Message) ([]byte, error) {
+	var size int32
+	attributes := int8(CompressionNoneCode)
+
+	// if compressing, replace the slice of messages with a single compressed
+	// message set.
+	if codec != nil {
+		var err error
+		if msgs, err = compress(codec, msgs...); err != nil {
+			return nil, err
+		}
+		attributes = codec.Code()
+	}
 
 	for _, msg := range msgs {
-		writeMessage(w, msg.Offset, attributes, msg.Time, msg.Key, msg.Value)
+		size += 8 + // offset
+			4 + // message size
+			4 + // crc
+			1 + // magic byte
+			1 + // attributes
+			8 + // timestamp
+			sizeofBytes(msg.Key) +
+			sizeofBytes(msg.Value)
 	}
 
-	return w.Flush()
+	buf := &bytes.Buffer{}
+	buf.Grow(int(size))
+	bufWriter := bufio.NewWriter(buf)
+
+	for _, msg := range msgs {
+		writeMessage(bufWriter, msg.Offset, attributes, msg.Time, msg.Key, msg.Value)
+	}
+	bufWriter.Flush()
+
+	return buf.Bytes(), nil
+}
+
+func writeRecordBatch(codec CompressionCodec, correlationID int32, clientId, topic string, partition int32, timeout time.Duration, requiredAcks int16, msgs ...Message) ([]byte, error) {
+	var size int32
+	size = 8 + // base offset
+		4 + // batch length
+		4 + // partition leader epoch
+		1 + // magic
+		4 + // crc
+		2 + // attributes
+		4 + // last offset delta
+		8 + // first timestamp
+		8 + // max timestamp
+		8 + // producer id
+		2 + // producer epoch
+		4 // base sequence
+
+	for _, msg := range msgs {
+		size += estimatedRecordSize(&msg)
+	}
+
+	buf := &bytes.Buffer{}
+	buf.Grow(int(size))
+	bufWriter := bufio.NewWriter(buf)
+
+	baseTime := msgs[0].Time
+
+	baseOffset := baseOffset(msgs...)
+
+	writeInt64(bufWriter, baseOffset)
+
+	remainderBuf := &bytes.Buffer{}
+	remainderBuf.Grow(int(size - 12)) // 12 = batch length + base offset sizes
+	remainderWriter := bufio.NewWriter(remainderBuf)
+
+	writeInt32(remainderWriter, -1) // partition leader epoch
+	writeInt8(remainderWriter, 2)   // magic byte
+
+	//writeInt16(remainderWriter, 0) // TODO write crc
```

Comment:
Is this expected to remain commented out?

Reply:
It should be removed. I calculate crc below.